### PR TITLE
[FIRRTL] Keep the Annotation Target Caches updated

### PR DESCRIPTION
**Pull request by prithayan**
Add helpers to keep AnnoTarget caches updated.

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h [hunk 1]
```diff
@@ -84,6 +84,33 @@ struct AnnoTargetCache {
     return targets.lookup(name);
   }
 
+  /// Replace `oldOp` with `newOp` in the target cache. The new and old ops can
+  /// have different names.
+  void replaceOp(Operation *oldOp, Operation *newOp) {
+    auto namOp = dyn_cast<FNamableOp>(oldOp);
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h:90]: Requiring these to be namable is out-of-sync with the code that populates this cache (`gatherTargets`).  A quick review shows the following are considered for annotation targets but would be ignored here: `MemoryPortOp` and `PrintFOp`.

Not sure if perhaps those should be `FNamable` or the "is nameable target considered for annotations" should be captured in a method somewhere, but it seems creation and updating should have same idea here.

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h [hunk 2]
```diff
@@ -84,6 +84,33 @@ struct AnnoTargetCache {
...
+      return;
+    if (auto name = namOp.getNameAttr(); name && !name.getValue().empty())
+      targets.erase(name);
+    if (auto name = cast<FNamableOp>(newOp).getNameAttr();
+        name && !name.getValue().empty())
+      targets.insert({name, OpAnnoTarget(newOp)});
+  }
+
+  /// Add a new module port to the target cache.
+  void insertPorts(FModuleLike mod, size_t portNo) {
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h:101]:
```suggestion
  void insertPort(FModuleLike mod, size_t portNo) {
```
Only inserts a single port.

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h [hunk 3]
```diff
@@ -107,6 +134,36 @@ struct CircuitTargetCache {
...
+  void replaceOp(Operation *oldOp, Operation *newOp) {
+    auto mod = newOp->getParentOfType<FModuleOp>();
+    auto it = targetCaches.find(mod);
+    if (it == targetCaches.end())
+      return;
+    it->getSecond().replaceOp(oldOp, newOp);
+  }
+
+  /// Add a new module port to the target cache.
+  void insertPorts(FModuleLike mod, size_t portNo) {
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h:151]:
```suggestion
  void insertPort(FModuleLike mod, size_t portNo) {
```

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h [hunk 4]
```diff
@@ -84,6 +84,33 @@ struct AnnoTargetCache {
...
 
+  /// Replace `oldOp` with `newOp` in the target cache. The new and old ops can
+  /// have different names.
+  void replaceOp(Operation *oldOp, Operation *newOp) {
+    auto namOp = dyn_cast<FNamableOp>(oldOp);
+    if (!namOp)
+      return;
+    if (auto name = namOp.getNameAttr(); name && !name.getValue().empty())
+      targets.erase(name);
+    if (auto name = cast<FNamableOp>(newOp).getNameAttr();
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h:95]: I don't know the use-case, but presumably this is for keeping the annotation target cache up to date while adding/replacing operations-- is it guaranteed the nameable operations are only replaced with other nameable operations? (re:`cast`).

**Comment by prithayan** [comment]
Thanks @dtzSiFive for the review. Merged this.

**Pull request merged**